include/cuco/detail/hash_functions/identityhash.cuh

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2023-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/detail/hash_functions/utils.cuh>
+
+namespace cuco::detail {
+
+/**
+ * @brief An Identity hash function to hash the given argument on host and device
+.*
+ * @note IdentityHash is perfect iff hash_table_capacity >= |input set|
+ *
+ * @note IdentityHash is only intended to be used perfectly.
+ *
+ * @note Perfect hashes are deterministic, and thus do not need seeds.
+ *
+ * -----------------------------------------------------------------------------
+ *
+ * @tparam Key The type of the values to hash
+ */
+template <typename Key>
+struct IdentityHash {
+  using argument_type = Key;  ///< The type of the values taken as argument
+  using result_type   = Key;  ///< The type of the hash values produced
+
+  /**
+   * @brief Returns a hash value for its argument, as a value of type `result_type`.
+   *
+   * @param key The input argument to hash
+   * @return The resulting hash value for `key`
+   */
+  constexpr result_type __host__ __device__ operator()(Key const& key) const noexcept
+  {
+    return idfunct_(key);
+  }
+
+ private:
+  thrust::identity<Key> idfunct_;
+};  // Identity hash
```

Comment:
```cuda
template<typename Key>
struct identity_hash : public thrust::identity<Key> {
  using argument_type = Key;  ///< The type of the values taken as argument
  using result_type   = Key;  ///< The type of the hash values produced
};
```
This probably also works (and simplifies the code a bit).


Murmurhash and XXhash use PascalCase since they are ported from existing libraries. Identity hashers don't have such copyright issues thus we could use [snake_case](https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#code-and-documentation-style-and-formatting) directly for class naming.

Reply:
I agree, an alias typedef is the right way to go here since Thrust already provides exactly what we need.

Reply:
I vaguely recall @esoha-nvidia has a reason to not use the alias directly

Reply:
The issue comes from a future addition-- we likely want additional methods in all perfect hashers to help protect users from improper usage. @esoha-nvidia has a good explanation [here](https://github.com/NVIDIA/cuCollections/pull/505#discussion_r1646605810): 
>>> What we'll do is put a method in the perfect hashers (such as `SigbitHash` and `IdentityHash`) called `is_perfect` that returns `true`
I wrote the `IdentityHash` class out with these future additions in mind, but you are correct that in the present state, this file could be 5 lines long. I'm happy to make this change if it is still desired.

Reply:
Happy to use snake case!

Reply:
> we likely want additional methods in all perfect hashers to help protect users from improper usage.

Such methods should be at the probing scheme level instead of the hasher level. On the other hand, some cuco details may depend on `argument_type`/`result_type` thus using alias directly could break.


Let's move forward with the inheritance way:
```cuda
template<typename Key>
struct identity_hash : public thrust::identity<Key> {
  using argument_type = Key;  ///< The type of the values taken as argument
  using result_type   = Key;  ///< The type of the hash values produced
};
```